cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaSpecChecker.java

```diff
@@ -0,0 +1,139 @@
+                                            "log.message.format.version does not match the Kafka cluster version, which suggests that an upgrade is incomplete."));
+            }
+        }
+    }
+
+    /**
+     * Checks for a single-broker Kafka cluster using ephemeral storage. This is potentially a problem as it
+     * means any restarts of the broker will result in data loss, as the single broker won't allow for any
+     * topic replicas.
+     *
+     * @param warnings List to add a warning to, if appropriate.
+     */
+    private void checkKafkaStorage(List<Condition> warnings) {
+        if (kafkaCluster.getReplicas() == 1 && StorageUtils.usesEphemeral(kafkaCluster.getStorage())) {
+            warnings.add(buildCondition("KafkaStorage",
+                    "A Kafka cluster with a single replica and ephemeral storage will lose topic messages after any restart or rolling update."));
+        }
+    }
+
+    /**
+     * Checks for a single-node ZooKeeper cluster using ephemeral storage. This is potentially a problem as it
+     * means any restarts of the pod will cause the loss of cluster metadata.
+     *
+     * @param warnings List to add a warning to, if appropriate.
+     */
+    private void checkZooKeeperStorage(List<Condition> warnings) {
+        if (zkCluster.getReplicas() == 1 && StorageUtils.usesEphemeral(zkCluster.getStorage())) {
+            warnings.add(buildCondition("ZooKeeperStorage",
+                    "A ZooKeeper cluster with a single replica and ephemeral storage will be in a defective state after any restart or rolling update. It is recommended that a minimum of three replicas are used."));
+        }
+    }
+
+    /**
+     * Checks for an even number of ZooKeeper replicas. As ZooKeeper is dependent on maintaining a quorum,
+     * this means that users should deploy clusters with an odd number of nodes.
+     *
+     * @param warnings List to add a warning to, if appropriate.
+     */
+    private void checkZooKeeperReplicas(List<Condition> warnings) {
+        if (zkCluster.getReplicas() == 2) {
+            warnings.add(buildCondition("ZooKeeperReplicas",
+                    "Running ZooKeeper with two nodes is not advisable as both replicas will be needed to avoid downtime. It is recommended that a minimum of three replicas are used."));
+        } else if (zkCluster.getReplicas() % 2 == 0) {
+            warnings.add(buildCondition("ZooKeeperReplicas",
+                    "Running ZooKeeper with an odd number of replicas is recommended."));
+        }
+    }
+
+    private Condition buildCondition(String reason, String message) {
+        return new ConditionBuilder()
+                .withLastTransitionTime(timestamp)
+                .withType("Warning")
+                .withStatus("True")
+                .withReason(reason)
+                .withMessage(message)
+                .build();
+    }
+    private Date dateSupplier() {
+        return new Date();
+    }
```

Comment:
I was mostly thinking about passing the date supplier from the outside because it's created in the KafkaAssemblyOperator using the spec checker. I know they are the same (just returning `new Date()`) but if we do a change for any reason I don't have right now in the KafkaAssemblyOperator, it should be reflected in the spec checker. @tombentley wdyt? You had a related comment about this.

Reply:
Sorry if my comment wasn't clear. The point I was trying to make is that the KAO already uses a `Supplier<Date>` to abstract how it obtains the current time (in support of the "maintenance time" functionality). I think we should use the same `Supplier<Date>` instance here. In also makes it easier to write tests which depend on the "current" time.

Reply:
@tombentley I've tried moving it into ModelUtils alongside the formatter - was that what you had in mind?

Reply:
Almost. Please can we pass a `Supplier<Date>` to the constructor or `run()` method from the call site in KAO. Ideally only one place in KAO will depend on the ModelUtils method.